File: tests/DesktopGrass.Win2D.Tests/SimTests/EntitySkeletonTests.cs
// EntitySkeletonTests.cs - §13.2 roaming-entity subsystem.

using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class EntitySkeletonTests
{
    private const double Monitor1920 = 1920.0;

    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED,
                                double monitorWidth = Monitor1920,
                                double density = 1.0)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, monitorWidth, density),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(seed, monitorWidth);
        sim.ResetEntities(seed);
        return sim;
    }

    [Fact]
    public void EntityKindHasSpecLockedDiscriminants()
    {
        Assert.Equal(0, (int)EntityKind.None);
        Assert.Equal(1, (int)EntityKind.Tumbleweed);
        Assert.Equal(2, (int)EntityKind.Snowflake);
        Assert.Equal(3, (int)EntityKind.Sheep);
        Assert.Equal(4, (int)EntityKind.Cat);
        Assert.Equal(5, (int)EntityKind.Raindrop);
        Assert.Equal(64, Constants.MAX_ENTITIES_PER_MONITOR);
    }

    [Fact]
    public void SimDefaultsEntitiesToEmptyWithReservedCapacity()
    {
        var sim = new Sim();
        Assert.Empty(sim.Entities);
        Assert.True(sim.Entities.Capacity >= Constants.MAX_ENTITIES_PER_MONITOR);
    }

    [Fact]
    public void SetSceneClearsEntities()
    {
        var sim = BuildSim();
        sim.Entities.Add(new Entity
        {
            Kind = EntityKind.Tumbleweed,
            X = 100.0,
            Size = 10.0,
        });
        Assert.Single(sim.Entities);

        sim.SetScene(Scene.Grass);
        Assert.Empty(sim.Entities);
    }

    [Fact]
    public void TickEntitiesIsNoOpOnEmptyOutsideGrass()
    {
        var sim = BuildSim();
        sim.CurrentScene = Scene.Desert;
        ulong prngBefore = sim.AmbientPrng.State;
        int bladesBefore = sim.Blades.Length;

        sim.TickEntities(0.016);
        sim.TickEntities(0.5);

        Assert.Empty(sim.Entities);
        Assert.Equal(bladesBefore, sim.Blades.Length);
        Assert.Equal(prngBefore, sim.AmbientPrng.State);
    }

    [Fact]
    public void TickEntitiesAdvancesPopulatedEntity()
    {
        var sim = BuildSim();
        sim.CurrentScene = Scene.Desert;
        sim.Entities.Add(new Entity
        {
            Kind          = EntityKind.Tumbleweed,
            X             = 100.0,
            Y             = 50.0,
            Vx            = 50.0,
            Vy            = 0.0,
            Size          = 10.0,
            Rotation      = 0.5,
            RotationSpeed = 1.0,
            Age           = 0.0,
            Lifetime      = -1.0,
            Seed          = 0xDEADBEEF,
        });

        sim.TickEntities(0.5);

        Assert.Single(sim.Entities);
        var after = sim.Entities[0];
        Assert.Equal(100.0 + 50.0 * 0.5, after.X);
        Assert.Equal(50.0, after.Y);
        Assert.Equal(0.5 + 1.0 * 0.5, after.Rotation);
        Assert.Equal(0.5, after.Age);
        Assert.Equal(EntityKind.Tumbleweed, after.Kind);
    }

    [Fact]
    public void TickCallsTickEntitiesWiringCheck()
    {
        var sim = BuildSim();
        sim.CurrentScene = Scene.Desert;
        sim.Entities.Add(new Entity
        {
            Kind     = EntityKind.Snowflake,
            X        = 0.0,
            Y        = 0.0,
            Vx       = 10.0,
            Vy       = 20.0,
            Size     = 2.0,
            Age      = 0.0,
            Lifetime = 100.0,
        });

        sim.Tick(0.1, System.ReadOnlySpan<InputEvent>.Empty);
        Assert.Single(sim.Entities);
        Assert.Equal(1.0, sim.Entities[0].X, 9);
        Assert.Equal(2.0, sim.Entities[0].Y, 9);
    }
}

File: tests/DesktopGrass.Native.Tests/src/entity_skeleton_tests.cpp
// entity_skeleton_tests.cpp
//
// Entity subsystem skeleton tests (architecture.md §13.2).
//
// Coverage:
//   * EntityKind discriminants match the spec ({None=0, Tumbleweed=1,
//     Snowflake=2}).
//   * MAX_ENTITIES_PER_MONITOR is the locked cap (= 64).
//   * sim_init defaults sim.entities to empty, capacity >= cap.
//   * sim_set_scene clears entities (currently a no-op since the Grass
//     scene generates none; §14/§15 add per-scene generators).
//   * sim_tick_entities is safe on empty (no exceptions, no growth).
//   * Tick on empty entities does not perturb other sim state (blades
//     untouched, ambient PRNG untouched).

#include "TestHelpers.h"
#include "Sim.h"

using namespace desktopgrass;

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(EntitySkeletonTests)
{
public:
TEST_METHOD(EntityKindHasSpecLockedDiscriminants) {
    Assert::IsTrue(static_cast<int>(EntityKind::None)       == 0);
    Assert::IsTrue(static_cast<int>(EntityKind::Tumbleweed) == 1);
    Assert::IsTrue(static_cast<int>(EntityKind::Snowflake)  == 2);
    Assert::IsTrue(static_cast<int>(EntityKind::Sheep)      == 3);
    Assert::IsTrue(static_cast<int>(EntityKind::Cat)        == 4);
    Assert::IsTrue(static_cast<int>(EntityKind::Bunny)      == 6);
    Assert::IsTrue(static_cast<int>(EntityKind::Butterfly)  == 7);
    Assert::IsTrue(static_cast<int>(EntityKind::Firefly)    == 8);
    Assert::IsTrue(static_cast<int>(EntityKind::Bird)       == 9);
    Assert::IsTrue(MAX_ENTITIES_PER_MONITOR == 64);
}

TEST_METHOD(SimInitReservesEntitiesCapacity) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, DEFAULT_DENSITY);
    Assert::IsTrue(sim.entities.empty());
    Assert::IsTrue(sim.entities.capacity() >= static_cast<std::size_t>(MAX_ENTITIES_PER_MONITOR));
    Assert::IsTrue(sim.entitySeed == CANONICAL_TEST_SEED);
}

TEST_METHOD(SimSetSceneClearsEntities) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, 1.0);
    // Push a fake entity directly to verify scene-transition removal runs.
    Entity fake{};
    fake.kind = EntityKind::Tumbleweed;
    fake.x = 100.0;
    sim.entities.push_back(fake);
    Assert::IsTrue(sim.entities.size() == 1);

    sim_set_scene(sim, Scene::Winter);
    Assert::IsTrue(sim.entities.empty());
    Assert::IsTrue(sim.entities.capacity() >= static_cast<std::size_t>(MAX_ENTITIES_PER_MONITOR));
}

TEST_METHOD(SimTickEntitiesIsANoOpOnEmptyOutsideGrass) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, 1.0);
    sim.currentScene = Scene::Desert;
    const auto bladesBefore = sim.blades.size();
    const auto prngBefore   = sim.ambientPrng.state;

    sim_tick_entities(sim, 0.016);
    sim_tick_entities(sim, 0.5);

    Assert::IsTrue(sim.entities.empty());
    Assert::IsTrue(sim.blades.size() == bladesBefore);
    Assert::IsTrue(sim.ambientPrng.state == prngBefore);
}

TEST_METHOD(SimTickEntitiesAdvancesAPopulatedEntity) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, 1.0);
    sim.currentScene = Scene::Desert;
    Entity e{};
    e.kind          = EntityKind::Tumbleweed;
    e.x             = 100.0;
    e.y             = 50.0;
    e.vx            = 50.0;    // DIP/sec
    e.vy            = 0.0;
    e.size          = 10.0;
    e.rotation      = 0.5;
    e.rotationSpeed = 1.0;     // rad/sec
    e.age           = 0.0;
    e.lifetime      = -1.0;    // infinite
    e.seed          = 0xDEADBEEF;
    sim.entities.push_back(e);

    const double dt = 0.5;
    sim_tick_entities(sim, dt);

    Assert::IsTrue(sim.entities.size() == 1);
    const Entity& after = sim.entities[0];
    Assert::IsTrue(after.x == Near(100.0 + 50.0 * dt));
    Assert::IsTrue(after.y == Near(50.0));
    Assert::IsTrue(after.rotation == Near(0.5 + 1.0 * dt));
    Assert::IsTrue(after.age == Near(0.0 + dt));
    Assert::IsTrue(after.kind == EntityKind::Tumbleweed);
}

TEST_METHOD(SimTickCallsSimTickEntitiesWiringCheck) {
    Sim sim = sim_init(CANONICAL_TEST_SEED, 1920.0, 1.0);
    sim.currentScene = Scene::Desert;
    Entity e{};
    e.kind = EntityKind::Snowflake;
    e.x = 0.0; e.y = 0.0;
    e.vx = 10.0; e.vy = 20.0;
    e.size = 2.0;
    e.age = 0.0; e.lifetime = 100.0;
    sim.entities.push_back(e);

    sim_tick(sim, 0.1, nullptr, 0);
    Assert::IsTrue(sim.entities.size() == 1);
    Assert::IsTrue(sim.entities[0].x == Near(1.0));   // 10 * 0.1
    Assert::IsTrue(sim.entities[0].y == Near(2.0));   // 20 * 0.1
}
};
}
